### PR TITLE
Pecos 1D3D interactivity

### PR DESCRIPTION
**Pull request by bioinsilico**
- Using saguaro 1D3D module in pecos
- Adding logic for displaying MSA
- react 18
- update + audit
- webpack-dev-server

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -58,27 +55,34 @@
...
+    "sass-loader": "^13.2.0",
     "stream-browserify": "^3.0.0",
     "style-loader": "^3.3.1",
+    "ts-loader": "^9.4.2",
     "typescript": "^4.5.2",
+    "uniqid": "^5.4.0",
     "webpack": "^5.64.4",
-    "webpack-cli": "^4.9.2"
+    "webpack-cli": "^4.9.2",
+    "webpack-dev-server": "^4.11.1"
```
Comment (valasatava) [inline package.json:83]: with this there are now two ways to run a development server and the old way is broken, can you please refactor it to either fix and keep the old way or remove it completely

File: src/saguaro-3d/ExternalAlignmentProvider.ts [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+        this.alignment = alignment;
+        this.alignmentReference = alignmentReference;
+    }
+
+    async collect(requestConfig: AlignmentCollectConfig, filter?: Array<string>): Promise<AlignmentResponse> {
+        return new Promise((resolve)=>{
+            this.data().then((d)=>resolve(d));
+        });
+    }
+    async getTargets(): Promise<string[]> {
```
Comment (valasatava) [inline src/saguaro-3d/ExternalAlignmentProvider.ts:45]: could be made simple by using `await`

```
const d = await this.data();
return d.target_alignment?.map(ta=>ta?.target_id ?? 'NA') ?? [];
```

File: src/saguaro-3d/ExternalAlignmentProvider.ts [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    }
+    async getAlignmentLength(): Promise<number> {
+        return new Promise((resolve)=>{
+            this.data().then(d=>{
+                const ends = d.target_alignment?.map(ta=>ta?.aligned_regions?.[ta?.aligned_regions?.length - 1]?.query_end);
+                resolve(Math.max(...(ends as number[])));
+            });
+        });
+    }
+    async getAlignment(): Promise<AlignmentResponse> {
```
Comment (valasatava) [inline src/saguaro-3d/ExternalAlignmentProvider.ts:51]: maybe just `return this.data();`

File: src/saguaro-3d/ExternalAlignmentProvider.ts [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+            this.data().then((d)=>resolve(d));
+        });
+    }
+    async getTargets(): Promise<string[]> {
+        return new Promise((resolve)=>{
+            this.data().then((d)=>resolve(d.target_alignment?.map(ta=>ta?.target_id ?? 'NA') ?? []));
+        });
+    }
+    async getAlignmentLength(): Promise<number> {
+        return new Promise((resolve)=>{
```
Comment (valasatava) [inline src/saguaro-3d/ExternalAlignmentProvider.ts]: simpler with `await`

File: src/saguaro-3d/ExternalAlignmentProvider.ts [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    }
+    async getAlignment(): Promise<AlignmentResponse> {
+        return new Promise((resolve)=>{
+            this.data().then(d=>resolve(d));
+        });
+    }
+    private async data(): Promise<AlignmentResponse> {
+        if (this.alignmentResponse)
+            return this.alignmentResponse;
+        return new Promise((resolve)=>{
```
Comment (valasatava) [inline src/saguaro-3d/ExternalAlignmentProvider.ts]: `await` to the rescue

File: src/saguaro-3d/ExternalAlignmentProvider.ts [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    private readonly alignment: StructureAlignmentResponse;
+    private readonly alignmentReference: AlignmentReference;
+    constructor(alignment: StructureAlignmentResponse, alignmentReference: AlignmentReference) {
+        this.alignment = alignment;
+        this.alignmentReference = alignmentReference;
+    }
+
+    async collect(requestConfig: AlignmentCollectConfig, filter?: Array<string>): Promise<AlignmentResponse> {
+        return new Promise((resolve)=>{
+            this.data().then((d)=>resolve(d));
```
Comment (valasatava) [inline src/saguaro-3d/ExternalAlignmentProvider.ts]: could be made simple by using `await`

File: src/saguaro-3d/AlignmentTrackFactory.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
+import {
+    TrackFactoryInterface
+} from '@rcsb/rcsb-saguaro-app/build/dist/RcsbFvWeb/RcsbFvFactories/RcsbFvTrackFactory/TrackFactoryInterface';
+import {
+    AlignmentRequestContextType
+} from '@rcsb/rcsb-saguaro-app/build/dist/RcsbFvWeb/RcsbFvFactories/RcsbFvTrackFactory/TrackFactoryImpl/AlignmentTrackFactory';
+import { TargetAlignment } from '@rcsb/rcsb-api-tools/build/RcsbGraphQL/Types/Borrego/GqlTypes';
+import { RcsbFvDisplayTypes, RcsbFvRowConfigInterface } from '@rcsb/rcsb-saguaro';
```
Comment (valasatava) [inline src/saguaro-3d/AlignmentTrackFactory.ts:8]: this is missing with a fresh install

File: src/saguaro-3d/AlignmentReference.ts [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+                rcsbId: v.alignmentId,
+                sequence: missingSeq.find(s=>s.rcsbId === `${v.entryId}${TagDelimiter.instance}${v.instanceId}`)?.sequence ?? ''
+            }))
+        );
+    }
+
+    private addRef(id: string, alignment: AlignmentRefType, target: AlignmentRefType): void {
+        const map: AlignmentRefType = Array(this.alignmentRefMap.length).fill(undefined);
+        alignment.forEach((v, n)=>{
+            if (typeof v === 'undefined')
```
Comment (valasatava) [inline src/saguaro-3d/AlignmentReference.ts]: v === undefined should be OK

File: src/saguaro-3d/molstar-trajectory/AlignmentTrajectoryPresetProvider.ts [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        apply: async (trajectory: StateObjectRef<PluginStateObject.Molecule.Trajectory>, params: AlignmentTrajectoryParamsType, plugin: PluginContext) => {
+            if (!params.pdb)
+                return {};
+            const modelParams = { modelIndex: params.modelIndex || 0 };
+            const builder = plugin.builders.structure;
+
+            const model = await builder.createModel(trajectory, modelParams);
+            const modelProperties = await builder.insertModelProperties(model);
+            let structure;
+            let assemblyId = 1;
```
Comment (valasatava) [inline src/saguaro-3d/molstar-trajectory/AlignmentTrajectoryPresetProvider.ts]: you can find the assembly with a specific entity using

```
ModelSymmetry.Provider.get(model.data!)?.assemblies[0].operatorGroups[0].asymIds
```

and

```
model.data?.properties.structAsymMap.get('A')?.entity_id
```

File: src/saguaro-3d/molstar-trajectory/AlignmentRepresentationPresetProvider.ts [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+                    }
+                    const comp = await plugin.builders.structure.tryCreateComponentFromExpression(
+                        structureCell,
+                        MS.struct.generator.atomGroups({
+                            'chain-test': MS.core.logic.and([
+                                MS.core.rel.eq([MS.ammp('label_asym_id'), alignedAsymId]),
+                                MS.core.rel.eq([MS.acp('operatorName'), alignedOperatorName])
+                            ])
+                        }),
+                        uniqid(`${entryId}${TagDelimiter.entity}${alignedEntityId}${TagDelimiter.instance}${alignedAsymId}${TagDelimiter.entity}${alignedOperators.join(',')}`),
```
Comment (valasatava) [inline src/saguaro-3d/molstar-trajectory/AlignmentRepresentationPresetProvider.ts]: could use `structureCell.transform.ref` as a prefix to get a unique id instead of using the `uniqid` library

File: src/saguaro-3d/molstar-trajectory/AlignmentRepresentationPresetProvider.ts [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+    StructureElement,
+    StructureProperties as SP
+} from 'molstar/lib/mol-model/structure';
+
+import { MolScriptBuilder as MS } from 'molstar/lib/mol-script/language/builder';
+import uniqid from 'uniqid';
+import { createSelectionExpressions } from '@rcsb/rcsb-molstar/build/src/viewer/helpers/selection';
+import { ParamDefinition as PD } from 'molstar/lib/mol-util/param-definition';
+
+import { TagDelimiter } from '@rcsb/rcsb-saguaro-app';
```
Comment (valasatava) [inline src/saguaro-3d/molstar-trajectory/AlignmentRepresentationPresetProvider.ts]: would be better to have it in a more general library (api tools?)

File: src/saguaro-3d/molstar-trajectory/AlignmentTrajectoryPresetProvider.ts [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+                        }
+                    }
+                }
+                if (!entityCheck)
+                    plugin.managers.structure.hierarchy.remove([
+                        plugin.managers.structure.hierarchy.current.structures[plugin.managers.structure.hierarchy.current.structures.length - 1]
+                    ]);
+            } while (!entityCheck);
+            if (structure.data?.model.id)
+                colorConfig.setAlignmentIdToModel(structure.data?.model.id, alignmentId);
```
Comment (valasatava) [inline src/saguaro-3d/molstar-trajectory/AlignmentTrajectoryPresetProvider.ts]: instead of creating a TrajectoryHierarchyPresetProvider and color theme for each structure you can add data to structure.data.inheritedPropertyData like in https://github.com/molstar/rcsb-molstar/blob/b8665e2731dbdd3b04a37bfa495e824a7db92ea2/src/viewer/helpers/preset.ts#L189 and access that in a generic color theme

File: src/saguaro-3d/molstar-trajectory/AlignmentRepresentationPresetProvider.ts [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+                    alignedEntityId = SP.chain.label_entity_id(l);
+                    alignedAsymId = SP.chain.label_asym_id(l);
+                    alignedOperatorName = SP.unit.operator_name(l);
+                    alignedType = SP.entity.type(l);
+                    const alignedOperators: string[] = SP.unit.pdbx_struct_oper_list_ids(l);
+                    if (alignedOperators.length === 0) alignedOperators.push('0');
+                    if (alignedType !== 'polymer')
+                        return {};
+                    if (params.transform?.[0].transform) {
+                        await matrixAlign(plugin, structureRef, params.transform?.[0].transform);
```
Comment (valasatava) [inline src/saguaro-3d/molstar-trajectory/AlignmentRepresentationPresetProvider.ts]: would be more logical to have this in the TrajectoryHierarchyPresetProvider which creates structures

File: src/saguaro-3d/molstar-trajectory/AlignmentRepresentationPresetProvider.ts [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+                    representationMap['aligned'] = builder.buildRepresentation(update, comp, {
+                        color: CLOSE_RESIDUE_COLOR,
+                        type: 'cartoon'
+                    });
+
+                    await update.commit({ revertOnError: false });
+                    break;
+                }
+            }
+
```
Comment (valasatava) [inline src/saguaro-3d/molstar-trajectory/AlignmentRepresentationPresetProvider.ts]: I find it a bit hard to follow, for future readers it could be helpful to add some comments saying what component you are creating next. Also some helper functions to create the component labels/keys could be useful for readability.

File: src/saguaro-3d/molstar-trajectory/FlexibleAlignmentTrajectoryPresetProvider.ts [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+import { RootStructureDefinition } from 'molstar/lib/mol-plugin-state/helpers/root-structure';
+import { StateObject } from 'molstar/lib/mol-state/object';
+import { Structure, StructureElement, StructureProperties as SP } from 'molstar/lib/mol-model/structure';
+import { RigidTransformType } from '@rcsb/rcsb-saguaro-3d/lib/RcsbFvStructure/StructureUtils/StructureLoaderInterface';
+import { FlexibleAlignmentBuiltIn } from './FlexibleAlignmentBuiltIn';
+import { representationPresetProvider } from './FlexibleAlignmentRepresentationPresetProvider';
+import { AlignmentTrajectoryParamsType } from './AlignmentTrajectoryPresetProvider';
+import { ColorConfig } from '../ExternalAlignmentProvider';
+
+export function getTrajectoryPresetProvider(alignmentId: string, colorConfig: ColorConfig) {
```
Comment (valasatava) [inline src/saguaro-3d/molstar-trajectory/FlexibleAlignmentTrajectoryPresetProvider.ts]: the only difference to AlignmentTrajectoryPresetProvider seems to be the transform handling, I think it makes sence to merge the two files.

File: src/saguaro-3d/molstar-trajectory/Coloring.ts [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+export function closeResidueColorThemeProvider(colorConfig: ColorConfig): ColorTheme.Provider<{}, typeof CLOSE_RESIDUE_COLOR> {
+    function closeResidueColorTheme(ctx: ThemeDataContext, props: ParamDefinition.Values<{}>): ColorTheme<{}> {
+
+        const L = StructureElement.Location.create();
+        const locationColor = (location: StructureElement.Location) =>{
+            if (!ctx.structure)
+                return Color(0x777777);
+
+            const modelId = ctx.structure.model.id;
+            const closeResidues = colorConfig.getCloseResidues(modelId);
```
Comment (valasatava) [inline src/saguaro-3d/molstar-trajectory/Coloring.ts]: if the hierarchy preset would add ctx.structure.inheritedPropertyData.closeResidueConfig (or some other name) it could be used instead of creating many color themes

File: src/saguaro-3d/molstar-trajectory/FlexibleAlignmentBuiltIn.ts [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+                StructureElement.Location.set(l, structure, unit, unit.elements[0]);
+                if (SP.chain.label_entity_id(l) === entityId || SP.chain.label_asym_id(l) === instanceId) {
+                    alignedEntityId = SP.chain.label_entity_id(l);
+                    alignedAsymId = SP.chain.label_asym_id(l);
+                    alignedOperatorName = SP.unit.operator_name(l);
+                    alignedType = SP.entity.type(l);
+                    const alignedOperators: string[] = SP.unit.pdbx_struct_oper_list_ids(l);
+                    if (alignedOperators.length === 0) alignedOperators.push('0');
+                    if (alignedType !== 'polymer')
+                        throw new Error('');
```
Comment (valasatava) [inline src/saguaro-3d/molstar-trajectory/FlexibleAlignmentBuiltIn.ts]: should have an error message?

File: src/saguaro-3d/molstar-trajectory/FlexibleAlignmentBuiltIn.ts [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+                        const ts: Structure = Structure.transform(s, trans.transform as unknown as Mat4);
+                        const u = ts.units[0];
+                        builder.addUnit(u.kind, u.model, u.conformation.operator, u.elements, u.traits, u.invariantId);
+                    }
+                    builder.endChainGroup();
+                    const blockStructure = builder.getStructure();
+                    return new PSO.Molecule.Structure(blockStructure, { label: structure.label });
+                }
+            }
+            throw new Error('');
```
Comment (valasatava) [inline src/saguaro-3d/molstar-trajectory/FlexibleAlignmentBuiltIn.ts]: should have an error message

File: src/ui/plugin.tsx [hunk 19]
```diff
@@ -1,46 +1,137 @@
...
+            const structAlignResponse = props.ctx.state.data.response.state;
+            const alignmentReference = new AlignmentReference();
+            if (structAlignResponse.results) {
+                alignmentReference.init(structAlignResponse.results).then(()=>{
+                    const dataProvider: RcsbModuleDataProviderInterface = {
+                        alignments: {
+                            collector: new RcsbStructuralAlignmentProvider(structAlignResponse, alignmentReference),
+                            context: {
+                                queryId: 'structural-alignment',
+                                group: GroupReference.MatchingUniprotAccession,
```
Comment (valasatava) [inline src/ui/plugin.tsx]: what's the relevance of it in the context of pairwise alignment tool?

File: src/ui/plugin.tsx [hunk 20]
```diff
@@ -1,46 +1,137 @@
...
+                            ),
+                            additionalContent: (props) => <></>
+                        },
+                        additionalConfig: {
+                            boardConfig: {
+                                rowTitleWidth: 110,
+                                disableMenu: true
+                            },
+                            trackConfigModifier: {
+                                alignment: (ac, ta)=>{
```
Comment (valasatava) [inline src/ui/plugin.tsx]: unused parameters. To improve readability and the intention of your code can prefix with underscore to inform that even though this callback method takes two parameters, they are not needed

File: src/index.html [hunk 21]
```diff
@@ -1,23 +1,21 @@
...
-  </head>
-  <body>
-    <noscript>You need to enable JavaScript to run this app.</noscript>
-    <div
-      id="app-container"
-      style="
-        width: 1170px;
+<head>
+  <meta charset="UTF-8">
+  <title>RCSB PDB PFV UniProt Group</title>
```
Comment (valasatava) [inline src/index.html]: looks like a copy pase artifact

File: src/ui/plugin.tsx [hunk 22]
```diff
@@ -1,46 +1,137 @@
...
+import { RcsbStructuralAlignmentProvider } from '../saguaro-3d/ExternalAlignmentProvider';
+import { AlignmentReference } from '../saguaro-3d/AlignmentReference';
+import { Subscription } from 'rxjs';
+import { closeResidueColorThemeProvider } from '../saguaro-3d/molstar-trajectory/Coloring';
 
+let panel3D: RcsbFv3DAlignmentProvider;
 export function ApplicationContextContainer(props: {ctx: ApplicationContext}) {
 
     const [state, setState] = useState<Status>(props.ctx.state.events.status.getValue());
     useObservable<Status>(props.ctx.state.events.status, setState);
```
Comment (valasatava) [inline src/ui/plugin.tsx:16]: was there a reasone to use `useEffect` instead of conditional rendering?

File: src/ui/plugin.tsx [hunk 23]
```diff
@@ -1,46 +1,137 @@
...
+    useEffect(()=>{
```
Comment (valasatava) [inline src/ui/plugin.tsx]: I think, specifying dependency is the "react" way to control when it runs

File: src/ui/plugin.tsx [hunk 24]
```diff
@@ -1,46 +1,137 @@
...
+                        molstarProps: {
+                            showStructureSourceControls: false,
+                            showStrucmotifSubmitControls: false,
+                            showSuperpositionControls: false
+                        }
+                    });
+                    panel3D.render().then(()=>{
+                        panel3D.pluginCall(plugin=>{
+                            (panel3D as unknown as {downloadSubscription: Subscription}).downloadSubscription = props.ctx.state.events.download.subscribe((e) => exportHierarchy(plugin, { format: 'cif' }));
+                            // TODO Very ugly but it does the job
```
Comment (valasatava) [inline src/ui/plugin.tsx]: could you, please, change the comment so it's more helpful to the future readers? Like, "needed to show the tooltip when Mol* is in full screen mode but not when it's collapsed" or similar

File: src/saguaro-3d/molstar-trajectory/AlignmentTrajectoryPresetProvider.ts [hunk 25]
```diff
@@ -0,0 +1,100 @@
...
+
+export function getTrajectoryPresetProvider(alignmentId: string, colorConfig: ColorConfig) {
+    return TrajectoryHierarchyPresetProvider({
+        id: 'alignment-to-reference',
+        display: {
+            name: 'Alignment to Reference'
+        },
+        isApplicable: (trajectory: PluginStateObject.Molecule.Trajectory, plugin: PluginContext): boolean => true,
+        params: (trajectory: PluginStateObject.Molecule.Trajectory | undefined, plugin: PluginContext): ParamDefinition.For<AlignmentTrajectoryParamsType> => ({
+            pdb: PD.Value<{entryId: string;entityId: string;}|{entryId: string;instanceId: string;}|undefined>(undefined),
```
Comment (valasatava) [inline src/saguaro-3d/molstar-trajectory/AlignmentTrajectoryPresetProvider.ts]: why not make it required?

**Review by valasatava — CHANGES_REQUESTED** [review]
Thank you, Joan! Please, see my comments below.

My main concern is adding multiple saguaro and molstar dependencies:
```
    "@rcsb/rcsb-saguaro-3d": "^3.0.0-data-provider.29",
    "@rcsb/rcsb-saguaro-app": "^5.0.0-data-provider.20",
    "@rcsb/rcsb-molstar": "^2.5.3",
    "molstar": "^3.31.4",

``` 

It can result in multiple versions of the same library being loaded into the application causing issues (plus a disadvantage of increased bundle size). Perhaps, re-exporting things that are needed upstream or factoring them out into smaller reusable libs could help. For example, things that are used from molstar library can be re-exported via rcsb-molstar. Dependency on both rcsb-saguaro-3d and rcsb-saguaro-app should be also avoided.

P.s. there is a conflict with master branch.

File: package.json (hunk 1 above)
Comment (bioinsilico) [reply package.json:83]: Yeah, old way never worked for me. I have uninstalled concurrently and http-server. I am not sure if there are more dependencies for the old testing.

File: src/saguaro-3d/ExternalAlignmentProvider.ts (hunk 2 above)
Comment (bioinsilico) [reply src/saguaro-3d/ExternalAlignmentProvider.ts:45]: bye bye annoying promises

File: src/saguaro-3d/AlignmentTrackFactory.ts (hunk 7 above)
Comment (bioinsilico) [reply src/saguaro-3d/AlignmentTrackFactory.ts:8]: For reason rcsb-saguaro was not a dep

File: src/saguaro-3d/molstar-trajectory/AlignmentRepresentationPresetProvider.ts (hunk 11 above)
Comment (bioinsilico) [reply src/saguaro-3d/molstar-trajectory/AlignmentRepresentationPresetProvider.ts]: Good point. I will move this to rcsb-api-tools

File: src/saguaro-3d/molstar-trajectory/AlignmentTrajectoryPresetProvider.ts (hunk 12 above)
Comment (bioinsilico) [reply src/saguaro-3d/molstar-trajectory/AlignmentTrajectoryPresetProvider.ts]: ColorTheme was already unique. I think the code is a bit more difficult to read now, there is some kind of black magic how colors are propagated, but I guess now it follows molstar philosophy

File: src/saguaro-3d/molstar-trajectory/AlignmentRepresentationPresetProvider.ts (hunk 14 above)
Comment (bioinsilico) [reply src/saguaro-3d/molstar-trajectory/AlignmentRepresentationPresetProvider.ts]: Don't look at rcsb-molstar. I have refactored the code to make it more readable

File: src/saguaro-3d/molstar-trajectory/Coloring.ts (hunk 16 above)
Comment (bioinsilico) [reply src/saguaro-3d/molstar-trajectory/Coloring.ts]: IMO molstar black magic makes code more difficult to read. Done!

File: src/ui/plugin.tsx (hunk 22 above)
Comment (bioinsilico) [reply src/ui/plugin.tsx:16]: I moved the react prop update logic into structure component

**Comment by bioinsilico** [comment]
Size won't be a problem since all modules are imported as libraries and never the final bundle. About the multiple version it can be a problem. A lot of bad experiences with multiple Reacts or rxjs ... I don't think is a good idea to re-export thinks the npm way is using peerDependencies and define what versions should be "injected" in the dependency. I have not used this feature yet, but is something that we need. For noe, I woul keep version consistent among the projects.

File: src/saguaro-3d/molstar-trajectory/Coloring.ts [hunk 26]
```diff
@@ -0,0 +1,62 @@
...
+    return {
+        factory: closeResidueColorTheme,
+        granularity: 'group',
+        color,
+        props
+    };
+}
+export const CloseResidueColorThemeProvider = {
+    name: CLOSE_RESIDUE_COLOR,
+    label: 'Close Residue Color',
```
Comment (valasatava) [inline src/saguaro-3d/molstar-trajectory/Coloring.ts]: Could we use "Equivalent Residues" as a label? This is how we call it in the summary table/Documentation

**Comment by valasatava** [comment]
> Size won't be a problem since all modules are imported as libraries and never the final bundle. About the multiple version it can be a problem. A lot of bad experiences with multiple Reacts or rxjs ... I don't think is a good idea to re-export thinks the npm way is using peerDependencies and define what versions should be "injected" in the dependency. I have not used this feature yet, but is something that we need. For noe, I woul keep version consistent among the projects.

I like what you propose relating using peerDependencies, we should definitely explore this later

**Review by valasatava — APPROVED** [review]
Thank you for all adjustments!
